keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,569 @@
+
+The [APIServer validation](https://github.com/kubernetes/kubernetes/blob/b9ce4ac212d150212485fa29d62a2fbd783a57b0/pkg/apis/core/validation/validation.go#L4162) that disallows empty `ports` will be relaxed, when `allPorts` is set.
+
+The value of `allPorts` field can be toggled on supported service types.
+
+### Supported service types
+
+Setting this field will be supported for:
+
+* ServiceType=ClusterIP
+* ServiceType=LoadBalancer
+
+This field is not applicable to ExternalName services.
+
+NodePort services are not supported.
+A NodePort service accepts traffic on a given port on the Node, redirecting it to the specified targetPort of the service endpoints. Support AllPorts for a NodePort service means -
+traffic to any port on the node, will be forwarded to its endpoints on the same port. This could potentially break networking on the node, if traffic for, say, port 22 got forwarded to \<endpoint IP\>:22.
+
+Headless services are not supported either. Headless services do not have a ClusterIP, nor can they be LoadBalancer services. SRV records for endpoints with empty port array will not be created. Hence, supporting AllPorts on Headless services has little value.
+
+Setting `allPorts` to true is not supported on services that specify [ExternalIPs in the spec.](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/#ServiceSpec)
+
+### Usage
+
+In order to expose the entire port range on a supported service, a user needs to:
+1) Create a LoadBalancer/ClusterIP Service and set `allPorts` to True.
+2) Leave the `ports` field empty. Populating the ports array is invalid when setting `allPorts` to True.
+
+There will be no NodePort allocation for the service in this case. The only exception is the NodePort for HealthChecks for LoadBalancers using `ExternalTrafficPolicy: Local`.
+
+
+### Service Transitions
+
+Consider the following transitions:
+
+* Changing from ClusterIP to LoadBalancer type, or vice versa
+  Preserving `allPorts` value as well as toggling is supported.
+
+* Setting an ExternalIP on a ClusterIP that has `allPorts` true
+  `allPorts` value has to be unset.
+
+* Changing from ClusterIP/LoadBalancer to NodePort type
+  `allPorts` value has to be unset.
+ 
+* Changing from ClusterIP/LoadBalancer to NodePort type
+  Once the service has been changed to ClusterIP/LoadBalancer type, `allPorts` field can be set.
+ 
+ 
+Transitioning from non-headless to headless and vice versa are not permitted by the API today.
+
+
+### Life of a request
+
+1) A service with ClusterIP a.b.c.d is configured with `allPorts` set to true. This service has 2 endpoints - pod1, pod2.
+
+2) A client(1.2.3.4) sends a request to a.b.c.d:8888 - it is received on a cluster node where:
+
+  a) firewall rules allow it
+
+  b) kube-proxy iptables rules DNAT it to one of the backend pod IPs - p.q.r.s (pod1)
+
+  c) request is received on pod1 with source IP - 1.2.3.4 and destination p.q.r.s:8888
+
+  d) pod1 responds directly to 1.2.3.4.
```

Comment:
It seems that we can get rid of externalTrafficPolicy=local too and simplify this even more